Fix #4513: DataTable add globalFilterFunction

jepsar opened the pull request:
Fix #4513

jepsar committed:
Add globalFilterFunction to dataTable

melloware reviewed (approved):
I like this idea.

Rapster commented on src/main/java/org/primefaces/component/datatable/DataTableBase.java:
```diff
@@ -670,6 +670,14 @@ public void setRowEditMode(String rowEditMode) {
         getStateHelper().put(PropertyKeys.rowEditMode, rowEditMode);
     }
 
+    public javax.el.MethodExpression getGlobalFilterFunction() {
```
Use simple class name and not with packages please. Good fix though, simple and neat :wink:

jepsar replied:
Sorry about that.. I just did some copy pasting here.

https://github.com/primefaces/primefaces/blob/master/src/main/java/org/primefaces/component/column/ColumnBase.java#L265

jepsar committed:
Update DataTableBase.java

import javax.el.MethodExpression

jepsar commented on src/main/java/org/primefaces/component/datatable/feature/FilterFeature.java:
```diff
@@ -180,10 +178,18 @@ public void filter(FacesContext context, DataTable table, List<FilterMeta> filte
+            if (globalFilterFunction != null
+                    && (Boolean) globalFilterFunction.invoke(elContext, new Object[]{table.getRowData(), globalFilterValue, filterLocale})) {
+                filteredData.add(table.getRowData());
+                continue;
```
This has kept me thinking.. I think it's wrong. Like this you cannot combine the global filter with column filters. Let me revise this.

jepsar committed:
Resolved conflicts

jepsar committed:
globalFilterFunction + column filter allowed

You can now have a global filter using a function and further narrow the results using column filters

jepsar commented:
@melloware @Rapster [I've revised `FilterFeature.java`](https://github.com/primefaces/primefaces/pull/4522/commits/3c2cd9211beaf94753b85e6efe55c1f33319857e) as mentioned in [my review comment](https://github.com/primefaces/primefaces/pull/4522#pullrequestreview-203037035). Can you please have a look?

Also, I tried to get rid of the conflicts for this branch by [copying changes from the master to `DataTableBase.java`](https://github.com/primefaces/primefaces/pull/4522/commits/89d3a72c1939e845bbf7bf38a7f0d6eae0fa9604), but that did not work... The "Resolve conflicts" button is grey for me, so I wouldn't know what else I could do here.

tandraschko commented:
can you please fix conflicts?

jepsar commented:
@tandraschko I wasn't aware at that time that I should create a branch within a fork.. that got me kind of stuck. Shall I provide a new PR without conflicts instead?

melloware commented:
Yeah if that is easier for you just close this PR and do your branch in your fork.  I typically name my branch like PF4513 or whatever the ticket number is so I can keep my branches straight!

jepsar commented:
@melloware Okay, I'll provide a new PR. By that way, I'll also provide a second one for PFE TimeAgo somewhere this weekend.

jepsar commented:
@tandraschko Please use PR #4674 instead of this one
